### PR TITLE
Top level unit tests for CostCalculator (v2)

### PR DESCRIPTION
**Pull request by kokosing**
Supersedes https://github.com/Teradata/presto/pull/525

### CODE REVIEW

File: presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        String queryId = runner.executeWithQueryId(runner.getDefaultSession(), query).getQueryId();
+        QueryPlan queryPlan = runner.getQueryPlan(new QueryId(queryId));
+
+        StageInfo stageInfo = runner.getQueryInfo(new QueryId(queryId)).getOutputStage().get();
+        return new MetricComparator().getMetricComparisons(queryPlan, stageInfo);
+    }
+
+    public static class Result
+    {
+        private final List<MetricComparison> metricComparison;
```
Comment (losipiuk) [inline presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java]: `metricComparisons`

File: presto-tests/src/test/java/com/facebook/presto/tests/TestDisplayTpchDistributedStats.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * This is a development tool for manual inspection of differences between
+     * cost estimates and actual execution costs. Its outputs need to be inspected
+     * manually because at this point no sensible assertions can be formulated
+     * for the entirety of TPCH queries.
+     */
+    @Test
+    void testCostEstimatesVsRealityDifferences()
+    {
+        IntStream.rangeClosed(1, NUMBER_OF_TPCH_QUERIES)
```
Comment (losipiuk) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestDisplayTpchDistributedStats.java:67]: static import

File: presto-tests/src/test/java/com/facebook/presto/tests/TestDisplayTpchDistributedStats.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                .filter(i -> i != 15) //query 15 creates a view, which TPCH connector does not support.
+                .forEach(i -> summarizeQuery(i, getTpchQuery(i)));
+    }
+
+    private String getTpchQuery(int i)
+    {
+        try {
+            String queryClassPath = "/io/airlift/tpch/queries/q" + i + ".sql";
+            return Resources.toString(getClass().getResource(queryClassPath), Charset.defaultCharset());
```
Comment (losipiuk) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestDisplayTpchDistributedStats.java:76]: UTF-8 would probably be more portable.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestDisplayTpchDistributedStats.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        }
+        catch (IOException e) {
+            throw Throwables.propagate(e);
```
Comment (losipiuk) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestDisplayTpchDistributedStats.java:79]: deprecated

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java [hunk 5]
```diff
@@ -126,7 +124,7 @@ private void summarizeQuery(int queryNumber, String query)
                         .collect(groupingBy(MetricComparison::getMetric, groupingBy(MetricComparison::result)));
 
         metricSummaries.forEach((metricName, resultSummaries) -> {
-            System.out.println(format("Summary for metric [%s]", metricName));
+            System.out.println(format("Summary for metric [%s] contains [%s] metrics", metricName, resultSummaries.size()));
```
Comment (ArturGajowy) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java]: contains [%s] results

Comment (ArturGajowy) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java]: `resultSummaries.size()` is always `4` - the number of `Result` enum values. What you're after is sth like `resultSummaries.values().stream().sum(List::size)`

File: presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    public Result check(@Language("SQL") String query)
+    {
+        return new Result(metricComparisons(query));
+    }
+
+    public List<MetricComparison> metricComparisons(@Language("SQL") String query)
+    {
+        String queryId = runner.executeWithQueryId(runner.getDefaultSession(), query).getQueryId();
+        QueryPlan queryPlan = runner.getQueryPlan(new QueryId(queryId));
+
```
Comment (ArturGajowy) [inline presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java:57]: remove blank line

File: presto-tests/src/test/java/com/facebook/presto/tests/TestDisplayTpchDistributedStats.java (hunk 2 above)
Comment (ArturGajowy) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestDisplayTpchDistributedStats.java:67]: nit: static import

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java [hunk 7]
```diff
@@ -36,8 +42,87 @@ public TestTpchDistributedStats()
...
+
+    @Test
+    void testFilter()
+    {
+        String query = "" +
+                "SELECT * " +
+                "FROM lineitem " +
+                "WHERE l_shipdate <= DATE '1998-12-01' - INTERVAL '90' DAY";
+        statisticsAssertion.check(query)
+                .outputHas(Metric.OUTPUT_ROW_COUNT, relativeError(Range.closed(-.55, -.45)));
```
Comment (ArturGajowy) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java:59]: add static import for Metric.*

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java [hunk 8]
```diff
@@ -36,8 +42,87 @@ public TestTpchDistributedStats()
...
+    }
+
+    @Test
+    void testJoin()
+    {
+        statisticsAssertion.check("SELECT * FROM  part, partsupp WHERE p_partkey = ps_partkey")
+                .outputHas(Metric.OUTPUT_ROW_COUNT, relativeError(Range.closed(.95, 1.05)));
```
Comment (ArturGajowy) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java]: I'd abstract from the output node and talk about the whole query instead. So please rename:
`outputHas` to `hasEstimateFor`. So the test looks like:
```java
     statisticsAssertion.check("SELECT * FROM  part, partsupp WHERE p_partkey = ps_partkey")
                .hasEstimateFor(OUTPUT_ROW_COUNT, relativeError(Range.closed(.95, 1.05)));
```

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java [hunk 9]
```diff
@@ -36,8 +42,87 @@ public TestTpchDistributedStats()
...
+    }
+
+    @Test
+    void testSetOperationsj()
+    {
+        statisticsAssertion.check("SELECT * FROM nation UNION SELECT * FROM nation")
+                .outputHasNoEstimate(Metric.OUTPUT_ROW_COUNT);
```
Comment (ArturGajowy) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java]: ditto: drop the `output` part from the method name. `hasNoEstimateFor`?

File: presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java [hunk 10]
```diff
@@ -53,19 +61,56 @@ public Result check(@Language("SQL") String query)
...
+                    0,
+                    String.format("Following metrics do not match: " + differingMetricComparisons));
+            return this;
+        }
+
+        public Result outputHasNoEstimate(Metric metric)
         {
-            Preconditions.checkArgument(metricComparison.isEmpty(), "No metric ");
-            this.metricComparison = ImmutableList.copyOf(metricComparison);
+            // TODO: remove passing strategy here
```
Comment (ArturGajowy) [inline presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java]: remove the TODO here and below

File: presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        {
+            return match(IS_OUTPUT_NODE, metric, strategy);
+        }
+
+        private Result match(Predicate<PlanNode> planNodeFilter, Metric metric, MetricComparisonStrategy strategy)
+        {
+            return testMetrics(planNodeFilter, metric, metricComparison -> metricComparison.result(strategy) == MATCH);
+        }
+
+        private Result testMetrics(Predicate<PlanNode> planNodeFilter, Metric metric, Predicate<MetricComparison> assertCondition)
```
Comment (ArturGajowy) [inline presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java:89]: please remove the planNodeFilter and always use the output node for comparison. There's no other use case until we have stable node IDs, and asserting about all the plan nodes is brittle

**Review by losipiuk — APPROVED** [review]

File: presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        {
+            List<MetricComparison> differingMetricComparisons = metricComparisons.stream()
+                    .filter(metricComparison -> metricComparison.getMetric() == metric)
+                    .filter(metricComparison -> planNodeFilter.test(metricComparison.getPlanNode()))
```
Comment (ArturGajowy) [inline presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java:93]: this method should raise an assertion error if there's not a result in either of the two lines above. In other words, metric comparisons should contain an element for all metrics, for all nodes.

Alternatively: maybe we should introduce a `MetricComparisons` class to guard this property?

I think this adds weight to the previous comment (remove `planNodeFilter`) as applying it only leaves us with the assumption, that there's a MetricComparison for all metrics, for the root node only)

File: presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+        public Result(List<MetricComparison> metricComparisons)
+        {
+            checkArgument(!metricComparisons.isEmpty(), "No metric comparisons given");
+            this.metricComparisons = ImmutableList.copyOf(metricComparisons);
+        }
+
+        public Result matches()
+        {
+            return match(alwaysTrue(), Metric.OUTPUT_ROW_COUNT, defaultTolerance());
```
Comment (ArturGajowy) [inline presto-tests/src/main/java/com/facebook/presto/tests/statistics/StatisticsAssertion.java:76]: let's not make OUTPUT_ROW_COUNT special in the API. Please replace the only usage with:

`statisticsAssertion.check(query).hasEstimateFor(OUTPUT_ROW_COUNT, noError())`

File: presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java [hunk 14]
```diff
@@ -13,144 +13,115 @@
...
-        IntStream.rangeClosed(1, NUMBER_OF_TPCH_QUERIES)
-                .filter(i -> i != 15) //query 15 creates a view, which TPCH connector does not support.
-                .forEach(i -> summarizeQuery(i, getTpchQuery(i)));
+        statisticsAssertion.check("SELECT * FROM  part, partsupp WHERE p_partkey = ps_partkey")
+                .outputHas(Metric.OUTPUT_ROW_COUNT, relativeError(Range.closed(.95, 1.05)));
     }
 
-    private String getTpchQuery(int i)
+    @Test
+    void testSetOperationsj()
```
Comment (ArturGajowy) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java:70]: typo

**Review by ArturGajowy — CHANGES_REQUESTED** [review]
looks good, several minors spotted

**Comment by ArturGajowy** [comment]
👍  for starting it against the epic branch :)

**Comment by kokosing** [comment]
Superseded by #533

**Pull request closed without merging**